[Regression] JUnit 5/6: Arquillian tests report errors instead of failures

Arquillian tests report all assertion failures as errors:
```
[INFO] --- failsafe:3.5.4:integration-test (integration-test) @ arq-error ---
[INFO] Tests will run in random order. To reproduce ordering use flag -Dfailsafe.runOrder.random.seed=790998900429250
[INFO] Using auto detected provider org.apache.maven.surefire.junitplatform.JUnitPlatformProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running com.reproducers.arq.err.its.StarterIT
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.177 s <<< FAILURE! -- in com.reproducers.arq.err.its.StarterIT
[ERROR] com.reproducers.arq.err.its.StarterIT.sanityCheck -- Time elapsed: 0.149 s <<< ERROR!
IdentifiedTestException [collectedExceptions={[engine:junit-jupiter]/[class:com.reproducers.arq.err.its.StarterIT]/[method:sanityCheck()]=org.opentest4j.AssertionFailedError: 
Expecting value to be true but was false}]
... stack trace ...
```

This is wrong. Unexpected failures and errors should be separated, so failsafe plugin reports them correctly.

According to ChatGPT, there is a minor difference between errors and failures:
> In JUnit Platform (JUnit 5), both "failed" and "errored" tests are represented by the TestExecutionResult.Status.FAILED status. JUnit does not distinguish between assertion failures and unexpected exceptions at the engine level—both are considered "failed".
To differentiate them in an extension, inspect the Throwable in TestExecutionResult.getThrowable():
If the throwable is an AssertionError, it's a test failure (assertion failed).
If it's any other exception, it's an error (unexpected exception).
Example:
```
if (result.getStatus() == TestExecutionResult.Status.FAILED) {
    Throwable throwable = result.getThrowable().orElse(null);
    if (throwable instanceof AssertionError) {
        // Test failed (assertion failed)
    } else {
        // Test errored (unexpected exception)
    }
}
```
This approach lets you distinguish between assertion failures and other errors in your JUnit extension.

--- a/junit5/core/src/main/java/org/jboss/arquillian/junit5/ArquillianExtension.java
+++ b/junit5/core/src/main/java/org/jboss/arquillian/junit5/ArquillianExtension.java
@@ -1,6 +1,8 @@
 package org.jboss.arquillian.junit5;
 
 import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.Optional;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Predicate;
 
@@ -141,13 +143,7 @@ public class ArquillianExtension implements BeforeAllCallback, AfterAllCallback,
                     result = interceptInvocation(invocation, extensionContext);
                 }
             }
-            if (result != null && result.getStatus() != TestResult.Status.PASSED) {
-                if(result.getThrowable() != null) {
-                    throw result.getThrowable();
-                } else {
-                    throw new TestAbortedException(result.getDescription());
-                }
-            }
+            throwError(result);
         }
     }
 
@@ -165,13 +161,7 @@ public class ArquillianExtension implements BeforeAllCallback, AfterAllCallback,
             invocation.proceed();
         } else {
             TestResult result = interceptInvocation(invocation, extensionContext);
-            if (result.getStatus() != TestResult.Status.PASSED) {
-                if(result.getThrowable() != null) {
-                    throw result.getThrowable();
-                } else {
-                    throw new TestAbortedException(result.getDescription());
-                }
-            }
+            throwError(result);
         }
     }
 
@@ -363,5 +353,36 @@ public class ArquillianExtension implements BeforeAllCallback, AfterAllCallback,
             return new ParameterResolutionException(msg);
         }
         return new ParameterResolutionException(msg, cause);
+    }
+
+    private static void throwError(final TestResult result) throws Throwable {
+        if (result == null || result.getStatus() == TestResult.Status.PASSED) {
+            return;
+        }
+        if (result.getStatus() == TestResult.Status.SKIPPED) {
+            throw new TestAbortedException(result.getDescription());
+        }
+        if (result.getStatus() == TestResult.Status.FAILED) {
+            throw failedAssertion(result.getThrowable()).orElseGet(() -> {
+                if (result.getThrowable() != null) {
+                    return result.getThrowable();
+                }
+                return new AssertionError(result.getDescription());
+            });
+        }
+        // We should not reach this, but it's a safeguard in case a new status is ever added and missed here
+        throw new AssertionError(result.getDescription());
+    }
+
+    private static Optional<Throwable> failedAssertion(Throwable exception) {
+        if (exception instanceof IdentifiedTestException) {
+            IdentifiedTestException identified = (IdentifiedTestException) exception;
+            Collection<Throwable> exceptions = identified.getCollectedExceptions().values();
+            if (exceptions.stream().allMatch(AssertionError.class::isInstance)) {
+                return exceptions.stream().filter(AssertionError.class::isInstance)
+                    .findFirst();
+            }
+        }
+        return Optional.empty();
     }
 }

